<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
                SELECT COUNT(*)
                FROM vacancies
'''

In [5]:
# результат запроса
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\1114379196.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


count
0  49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
                SELECT COUNT(*)
                FROM employers
'''

In [7]:
# результат запроса
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\3062646138.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


count
0  23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
                SELECT COUNT(*)
                FROM areas
'''

In [9]:
# результат запроса
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\928149064.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


count
0   1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
                SELECT COUNT(*)
                FROM industries
'''

In [11]:
# результат запроса
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2101691108.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


count
0    294

***

# Выводы по предварительному анализу данных

По результатам предварительного анализа данных мы узнали следующее:

1. Общее количество вакансий: **49,197**

2. Общее количество работодателей: **23,501**

3. Общее количество регионов: **1,362**

4. Общее количество сфер деятельности: **294**

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
                SELECT 
                    COUNT(*) cnt,
                    a.name
                FROM vacancies v
                JOIN areas a ON v.area_id = a.id
                GROUP BY 2
                ORDER BY 1 DESC
                LIMIT 5
'''

In [13]:
# результат запроса
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2457173557.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             name
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
                SELECT COUNT(id)
                FROM vacancies v
                WHERE salary_from IS NOT NULL
                OR salary_to IS NOT NULL
'''

In [15]:
# результат запроса
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\970655580.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса
query_4_3 = f'''
                SELECT
                    ROUND(AVG(salary_from)) rnd_salary_from,
                    ROUND(AVG(salary_to)) rnd_salary_to
                FROM vacancies
'''

In [17]:
# результат запроса
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2500607024.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


rnd_salary_from  rnd_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
                SELECT
                    schedule,
                    employment,
                    COUNT(*)
                FROM vacancies
                GROUP BY schedule, employment
                ORDER BY 3 DESC
'''

In [19]:
# результат запроса
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2022935265.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
                SELECT
                    experience,
                    COUNT(*)
                FROM vacancies
                GROUP BY 1
                ORDER BY 2 ASC
'''

In [21]:
# результат запроса
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2316260449.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

# Выводы по детальному анализу вакансий

* По результатам первого запроса мы получили топ 5 городов с наибольшим количеством вакансий. Наибольшее количество вакансий находится в **Москве**, т.к. это самый крупный и самый развитый город среди перечисленных. Следом за Москвой идут **Санкт-Петербург**, **Минск**, **Новосибирск** и **Алматы**.

* По результатам второго запроса мы получили статистику по заполнении поля с заработной платой. Из **49,197** вакансий, только в **24,073** заполнено хотя бы одно из двух полей.

* По результатам третьего запросы мы получили средние значения нижней и верхней границы заработной вилки. Нижняя граница состоявляет **71,065 тыс. руб**. Верхняя граница составляет **110,537 тыс. руб**.

* По результатам четвертого запросы мы узнали, что наиболее популярное сочетание типа рабочего графика и типа трудоустройства - это **Полный день - полная занятость**. Менее популярный, но все еще довольно часто встречающийся вариант - это **Удаленная работа - полная занятость**.

* По результатам пятого запроса мы выяснили, что больше половины вакансий ищут работников с опытом работы **от 1 года до 3 лет**.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
                SELECT
                    e.name,
                    COUNT(v.*)
                FROM vacancies v
                JOIN employers e ON v.employer_id = e.id
                GROUP BY e.name
                ORDER BY 2 DESC
                LIMIT 5
'''

In [23]:
# результат запроса
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\3834399874.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2 = f'''
                SELECT
                    a.name area_name,
                    COUNT(DISTINCT v.id) vacancies_cnt,
                    COUNT(DISTINCT e.id) employers_cnt
                FROM areas a
                LEFT JOIN vacancies v ON a.id = v.area_id
                LEFT JOIN employers e ON a.id = e.area
                WHERE v.id IS NULL
                GROUP BY a.id
                ORDER BY 3 DESC
                LIMIT 5
'''

In [25]:
# результат запроса
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2250345976.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


area_name  vacancies_cnt  employers_cnt
0              Россия              0            410
1           Казахстан              0            207
2  Московская область              0             75
3  Краснодарский край              0             19
4            Беларусь              0             18

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
# текст запроса
query_5_3 = f'''
                SELECT
                    e.name emp_name,
                    COUNT(DISTINCT v.area_id) vac_cnt
                FROM employers e
                JOIN vacancies v ON v.employer_id = e.id
                GROUP BY e.id
                ORDER BY 2 DESC
                LIMIT 5
'''

In [27]:
# результат запроса
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\201667002.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


emp_name  vac_cnt
0                  Яндекс      181
1              Ростелеком      152
2              Спецремонт      116
3  Поляков Денис Иванович       88
4               ООО ЕФИН        71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
# текст запроса
query_5_4 = f'''
                SELECT COUNT(e.id)
                FROM employers e
                LEFT JOIN employers_industries e_i ON e.id = e_i.employer_id
                WHERE e_i.industry_id IS NULL
'''

In [29]:
# результат запроса
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\3493105328.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
# текст запроса
query_5_5 = f'''
                SELECT
                    e.name emp_name,
                    COUNT(i_e.industry_id)
                FROM employers e
                LEFT JOIN employers_industries i_e ON e.id = i_e.employer_id
                GROUP BY e.id
                HAVING COUNT(i_e.industry_id) = 4
                ORDER BY 1
                OFFSET 2
                LIMIT 1
'''

In [31]:
# результат запроса
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\388711819.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


emp_name  count
0     2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
# текст запроса
query_5_6 = f'''
                SELECT COUNT(*)
                FROM employers_industries
                WHERE industry_id = '7.540'
'''

In [33]:
# результат запроса
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\821628714.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
# Сохраняем таблицу со списком городов-миллионников
table = pd.read_html(url)[1]
# Оставляем в таблице только столбец с названием городов
million_cities = tuple(table['Город'])

In [35]:
# текст запроса
query_5_7 = f'''
                WITH yandex_query AS
                (
                SELECT
                    a.name,
                    COUNT(a.id) cnt
                FROM vacancies v
                LEFT JOIN employers e ON v.employer_id = e.id
                LEFT JOIN areas a ON v.area_id = a.id
                WHERE e.name LIKE 'Яндекс' AND a.name IN {million_cities}
                GROUP BY a.id
                )
                SELECT *
                FROM yandex_query
                UNION
                SELECT
                    'Total',
                    SUM(cnt)
                FROM yandex_query
                ORDER BY 2 DESC
'''

In [36]:
# результат запроса
df = pd.read_sql_query(query_5_7, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\1965596796.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_7, connection)


name    cnt
0             Total  485.0
1            Москва   54.0
2   Санкт-Петербург   42.0
3      Екатеринбург   39.0
4   Нижний Новгород   36.0
5       Новосибирск   35.0
6           Воронеж   32.0
7         Краснодар   30.0
8            Самара   26.0
9               Уфа   26.0
10            Пермь   25.0
11           Казань   25.0
12   Ростов-на-Дону   25.0
13        Волгоград   24.0
14       Красноярск   23.0
15        Челябинск   22.0
16             Омск   21.0

***

# Выводы по анализу работодателей

1. По результатам первого запроса мы получили топ 5 работодателей по количеству вакансий. Лидирующую позицию удерживает **Яндекс** с **1933** вакансий.

2. По результатам второго запроса мы узнали, что у 410 работодателей указан регион **Россия**. При этом количество вакансий в этом регионе - 0.

3. По результатам третьего запроса мы получили топ 5 работодателей по регионам, в которых они публикуют свои вакансии. На первом месте находится **Яндекс**, что вполне логично т.к. эта компания опубликовала больше всего вакансий на сайте.

4. По результатам четвертого запроса мы узнали, что из **49,197** вакансий, в **8,419** не указана сфера деятельности. Это примерно **1/6** от всех вакансий.

5. По результатам пятого запроса мы получили компанию, которая находится на третьем месте в списке, отсортированном в алфавитном порядке, у которой указано 4 сферы деятельности. Этой компанией оказалась **2ГИС**.

6. По результатам шестого запроса мы узнали, что у **3,553** вакансий в качестве сферы деятельности указано **Разработка программного обеспечения**.

7. По результатам седьмого запроса мы получили информацию о количестве вакансий от компании **Яндекс** в различных городах-миллионниках. Больше всего вакансий опубликовано в **Москве** - **54** вакансии, **Санкт-Петербурге** - **42** вакансии и **Екатеринбурге** - **39** вакансий. Общее количество вакансий во всех городах-миллионниках - **485**.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [37]:
# текст запроса
query_6_1 = f'''
                SELECT COUNT(*)
                FROM vacancies
                WHERE LOWER(name) LIKE '%данн%'
                OR LOWER(name) LIKE '%data%'
'''

In [38]:
# результат запроса
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\366702258.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [39]:
# текст запроса
query_6_2 = f'''
                WITH ds_query AS
                (
                SELECT *
                FROM vacancies
                WHERE (name ILIKE '%data scientist%')
                OR (name ILIKE '%data science%')
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                )
                
                SELECT COUNT(id)
                FROM ds_query
                WHERE (name ILIKE '%junior%')
                OR (experience ILIKE '%Нет опыта%')
                OR (employment ILIKE '%Стажировка%')
'''

In [40]:
# результат запроса
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\2267038862.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


count
0     51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [41]:
# текст запроса
query_6_3 = f'''
                WITH ds_query AS
                (
                SELECT *
                FROM vacancies
                WHERE (name ILIKE '%data scientist%')
                OR (name ILIKE '%data science%')
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                )
                
                SELECT COUNT(id)
                FROM ds_query
                WHERE (key_skills ILIKE '%SQL%')
                OR (key_skills ILIKE '%postgres%')
'''

In [42]:
# результат запроса
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\3794299702.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


count
0    201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [43]:
# текст запроса
query_6_4 = f'''
                WITH ds_query AS
                (
                SELECT *
                FROM vacancies
                WHERE (name ILIKE '%data scientist%')
                OR (name ILIKE '%data science%')
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                )
                
                SELECT COUNT(id)
                FROM ds_query
                WHERE (key_skills ILIKE '%Python%')
'''

In [44]:
# результат запроса
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\4264665179.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


count
0    351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [45]:
# текст запроса
query_6_5 = f'''
                WITH ds_query AS
                (
                SELECT *
                FROM vacancies
                WHERE (name ILIKE '%data scientist%')
                OR (name ILIKE '%data science%')
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                ),
                
                key_skills_cnt AS
                (
                SELECT (LENGTH(key_skills) - LENGTH(REPLACE(key_skills, CHR(9), '')) + 1) str_len
                FROM ds_query
                )
                
                SELECT ROUND(AVG(str_len), 2)
                FROM key_skills_cnt
'''

In [46]:
# результат запроса
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\3654210517.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


round
0   6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [47]:
# текст запроса
query_6_6 = f'''
                WITH ds_query AS
                (
                SELECT *
                FROM vacancies
                WHERE (name ILIKE '%data scientist%')
                OR (name ILIKE '%data science%')
                OR (name ILIKE '%исследователь данных%')
                OR ((name LIKE '%ML%') AND (name NOT ILIKE '%HTML%'))
                OR (name ILIKE '%machine learning%')
                OR (name ILIKE '%машинн%обучен%')
                AND (key_skills IS NOT NULL)
                )
                
                SELECT
                    experience,
                    ROUND(AVG(COALESCE((salary_from + salary_to) / 2, salary_from, salary_to))) salary
                FROM ds_query
                GROUP BY experience
                
'''

In [48]:
# результат запроса
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Anatoly\AppData\Local\Temp\ipykernel_796\519194770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


experience    salary
0       От 3 до 6 лет  243115.0
1  От 1 года до 3 лет  145236.0
2         Более 6 лет       NaN
3           Нет опыта   74643.0

***

# Выводы по предметному анализу

1. По результатам первого запроса мы получили **1,771** вакансию, которые имеют отношение к данным.

2. По результатам второго запроса мы получили **51** вакансию, которые подходят для начинающих дата-сайентистов.

3. По результатам третьего запроса мы получили **201** вакансию, в которых, в качестве ключевых навыков, указан **SQL** или **postgres**.

4. По результатам четвертого запроса мы получили **351** вакансию, в которых, в качестве ключевых навыков, указан **Python**.

5. По результатам пятого запроса мы получили среднее значение количества ключевых навыков, которые указывают работодатели в вакансиях для DS. Это значение равно **6.41**.

6. По результатам шестого запроса мы получили среднюю заработную плату дата-сайентистов для каждого типа требуюмого опыта работы. Средняя заработная плата для вакансий **без опыта** составляет **74,643 тыс. руб**. В вакансиях с указанным опытом работы **От 1 года до 3 лет** средняя заработная плата составляет **145,236 тыс. руб**. При опыте работы **От 3 до 6 лет** средняя заработная плата составляет **243,115 тыс. руб**.

# Общий вывод по проекту

В результате нашей работы мы узнали следующее:

* В нашей базе данных **49,197** вакансий от **23,501** работодателя в **1,362** регионах поделенные на **294** различные сферы деятельности.

* Наибольшее количество вакансий опубликовано в **Москве**.

* Только у половины вакансий заполнено хотя бы одно из двух полей заработной платы. Среднее значение нижней и верхней значений равняются **71,065 тыс. руб.** и **110,537 тыс. руб.** соответственно.

* Наиболее популярным сочетанием рабочего графика и типа трудоустройства является **Полный день - Полная занятость**, а наиболее частым требованием к опыту работы является **от 1 года до 3 лет**.

* Примерно **17%** вакансий были опубликованы без указанаия сферы деятельности.

* У примерно **7%** вакансий в качестве сферы деятельности указана **Разработка программного обеспечения**.

* В нашей базе **1,771** вакансия, которые имеют отношения к данным. Из них всего **51** вакансия подходит для начинающих дата-сайентистов.

* Разница в средней заработной плате у сотрудников без опыта и сотрудников с опытом работы более 3 лет различается более чем в 3 раза.